File: backend/app/schemas/project.py
from typing import List, Optional

from app.models import project as model_project
from app.models import project_number as model_project_number
from app.schemas import base as schema_base
from app.schemas import company as schema_company
from app.schemas import user as schema_user


class ProjectGroup(schema_base.ORMBaseModel):
    rid: int
    name: str
    detail: str
    company: schema_company.Company


class ProjectGroupCreate(schema_base.ORMBaseModel):
    rid_companies: int
    name: str
    detail: str


class ProjectGroupUpdate(schema_base.ORMBaseModel):
    rid: int
    rid_companies: int
    name: str
    detail: str


class ProjectNumber(schema_base.ORMBaseModel):
    rid: int
    type: model_project_number.TypeNumber
    number: str
    note: str
    date_start: str
    date_end: str


class ProjectNumberCreate(schema_base.ORMBaseModel):
    rid_projects: int
    type: model_project_number.TypeNumber
    number: str
    note: str
    date_start: str
    date_end: str


class ProjectNumberUpdate(schema_base.ORMBaseModel):
    rid: int
    rid_projects: int
    type: model_project_number.TypeNumber
    number: str
    note: str
    date_start: str
    date_end: str


class TargetQuarter(schema_base.ORMBaseModel):
    year: int
    quarter: model_project.TypeQuarter


class SearchCondition(schema_base.ORMBaseModel):
    target: List[int]
    rid_users_pm: List[int]
    rid_users_pl: List[int]
    is_none_pre_approval: bool
    is_none_number_m: bool
    is_none_number_s: bool
    is_none_number_o: bool


class Project(schema_base.ORMBaseModel):
    rid: int
    project_group: ProjectGroup
    user_pm: schema_user.User
    user_pl: schema_user.User
    rank: model_project.TypeRank
    pre_approval: str
    name: str
    number_parent: str
    numbers: Optional[List[ProjectNumber]] = []
    number_m: bool
    number_s: bool
    number_o: bool
    amount_expected: int
    amount_order: int
    date_start: str
    date_delivery: str
    date_end: str
    target_quarter: int
    karte_plan: bool
    karte_report: bool
    checklist: bool


class ProjectCreate(schema_base.ORMBaseModel):
    rid_project_groups: int
    rid_users_pm: int
    rid_users_pl: int
    rank: int
    pre_approval: str
    name: str
    number_parent: str
    amount_expected: int
    amount_order: int
    date_start: str
    date_delivery: str
    date_end: str


class ProjectUpdate(schema_base.ORMBaseModel):
    rid: int
    rid_project_groups: int
    rid_users_pm: int
    rid_users_pl: int
    rank: int
    pre_approval: str
    name: str
    number_parent: str
    amount_expected: int
    amount_order: int
    date_start: str
    date_delivery: str
    date_end: str


class ProjectList(schema_base.ORMBaseModel):
    rid: int
    name: str
    company: schema_company.Company
    projects: List[Project]

File: frontend/app/src/types/Project.ts
import type { Company } from '@/types/Company'

import type { User } from '@/types/User'

// prettier-ignore
export enum TypeRank {
  NONE = 0,
  A    = 1,
  B    = 2,
  C    = 3,
  D    = 4,
  E    = 5
}

// prettier-ignore
export enum TypeNumber {
  NONE = 0,
  M    = 1,
  S    = 2,
  O    = 3
}

// prettier-ignore
export enum TypeQuarter {
  NONE = 0,
  Q1   = 1,
  Q2   = 2,
  Q3   = 3,
  Q4   = 4,
}

export interface ProjectGroup {
  rid: number
  name: string
  detail: string
  company: Company
}

export interface ProjectGroupCreate {
  rid_companies: number
  name: string
  detail: string
}

export interface ProjectGroupUpdate {
  rid: number
  rid_companies: number
  name: string
  detail: string
}

export interface ProjectNumber {
  rid: number
  type: TypeNumber
  number: string
  date_start: string
  date_end: string
}

export interface ProjectNumberCreate {
  rid_projects: number
  type: TypeNumber
  number: string
  date_start: string
  date_end: string
}

export interface ProjectNumberUpdate {
  rid: number
  rid_projects: number
  type: TypeNumber
  number: string
  date_start: string
  date_end: string
}

export interface Project {
  rid: number
  project_group: ProjectGroup
  user_pm: User
  user_pl: User
  rank: TypeRank
  pre_approval: string
  name: string
  number_parent: string
  numbers: ProjectNumber[]
  number_m: boolean
  number_s: boolean
  number_o: boolean
  amount_expected: number
  amount_order: number
  date_start: string
  date_delivery: string
  date_end: string
  karte_plan: boolean
  karte_report: boolean
  checklist: boolean
}

export interface ProjectCreate {
  rid_project_groups: number
  rid_users_pm: number
  rid_users_pl: number
  rank: TypeRank
  pre_approval: string
  name: string
  number_parent: string
  amount_expected: number
  amount_order: number
  date_start: string
  date_delivery: string
  date_end: string
}

export interface ProjectUpdate {
  rid: number
  rid_project_groups: number
  rid_users_pm: number
  rid_users_pl: number
  rank: TypeRank
  pre_approval: string
  name: string
  number_parent: string
  amount_expected: number
  amount_order: number
  date_start: string
  date_delivery: string
  date_end: string
}

export interface ProjectList {
  rid: number
  name: string
  company: Company
  projects: Project[]
}

export interface FilterQuarter {
  year: number
  quarter: TypeQuarter
}
